### ISSUE TITLE
How do I duck-type a class to behave like a DOMWidget?

### ISSUE THREAD

**Issue by suganya-suvadi** [jupyter-widgets/ipywidgets#1946]
**TLDR**
What is the recommended way to duck-type a class to behave like a DOMWidget?

**Details**
I have a class (`Foo`) that should be a custom `DOMWidget`; but, because of numerous miscellaneous problems from multiple inheritance, cannot inherit from `ipywidgets.DOMWidget`.

Rather, I have another class (`_WidgetBar`) that inherits from `DOMWidget` and am trying to delegate, from `Foo` to `_WidgetBar` whenever DOMWidget-like behavior is required out of `Foo`.

I am trying to get this done through duck-typing and have identified the following methods to be defined in the class - 

- `_widget_to_json`
- `_handle_displayed`

```
class Foo(parent):
   ....
   def __init__(self):
           self.widget_bar = _WidgetBar()

   def _widget_to_json(self):
          return self.widget_bar._widget_to_json()
   ....
```

Rather than handling all required methods like this on Foo directyly, is there an attribute/a method that a class should have for it to be accepted as a `DOMWidget` and through which all delegations can be handled in a single-go? So, Foo can have this method defined to be - 

```
def _as_dom_widget(self):
    return self.widget_bar
```
If not, is there a list of all methods that have to be defined in `Foo` for its instance to work like a `DOMWidget` under all circumstances?

**Comment by jasongrout** [comment]
Is it possible to create a DOMWidget subclass that holds an instance of Foo?

There are various places in the codebase where we use `isinstance` to check to see if something is a subclass of Widget. Things were really designed to work with subclasses, not interfaces.

**Comment by suganya-suvadi** [comment]
@jasongrout Thanks for the quick response!

I did consider doing the inverse - having `WidgetBar` (subclass of `DOMWidget`) hold an instance of `Foo`. The trouble here is the API of `Foo` is also quite extensive and has several user-facing methods. Most usage is going to be functionalities available on `Foo`, like - 
```
obj = Foo()
obj.do_stuff()
```
where `do_stuff` is defined on `Foo`.

DOMWidget-like behavior is expected on a few scenarios like display in Jupyter, or placing the instance inside `VBox`/`HBox`, embedding the widget in static contexts etc.

(Also, in case it sheds more light on the scenario, only `Foo` is public, `WidgetBar` is private. I've edited the original update to reflect this.)

Would you recommend any other way of doing this?

**Comment by jasongrout** [comment]
Perhaps Foo can have a `.widget` attribute that exposes a DOMWidget that can be displayed, inserted into a widget, etc. So your API is:

```
obj = Foo()
obj.do_stuff()
display(obj.widget)
HBox([obj.widget])
```

**Comment by suganya-suvadi** [comment]
This is the current scenario; this ticket was the last attempt before exposing the `.widget` attribute.

Thanks for all your input here!

**Comment by suganya-suvadi** [comment]
Closing this since there are no action items here; thanks!
